### DISCUSSION TITLE
Gaps in immutable programming support bring into question the soundness of the overall type system implementation

### DISCUSSION THREAD

**Discussion by mcandre** [dotnet/csharplang#8387] [General]
The lack of constant return types for a majority of standard library calls, and the lack of type inferencing altogether in constants, suggests the existence of latent bugs in the overall type system. For comparison, other languages like Haskell use explicit types as mainly a form of documentation for the user, inferring the type automatically and informing the developer in the event of a discrepancy between inferred vs explicit type information.

When will C# resolve these problems? Or are C# devs advised to migrate to Rust?

**Comment by CyrusNajmabadi** [comment]
>  for a majority of standard library calls,

The standard library is not part of C#.  Requests to them should be made at dotnet/runtime.

> and the lack of type inferencing altogether in constants

Not sure what is being referred to here.

> For comparison, other languages like Haskell use explicit types as mainly a form of documentation for the user

Sure. We've taken a different philosophy where we want your surface area to be explicit typed so that the code clearly indicates the API, and so that breaking ABI changes don't happen silently.  It's a philosophical difference.

> When will C# resolve these problems?

We do not consider this a problem.

> Or are C# devs advised to migrate to Rust?

You're welcome to use whatever tools you think best fit your needs.
